### ISSUE TITLE
TeamCity current build number

### ISSUE THREAD

**Issue by waynebrantley** [cake-build/cake#884]
Provider needs a way to determine current build number from TeamCity.
This is done through environment variable  "BUILD_NUMBER"

**Comment by devlead** [comment]
Agree would be great.

Looking thru the docs there're a few interesting environment variables

|Variable                                                    | Description                                         |
|----------------------------------------------------|------------------------------------------------|
|TEAMCITY_BUILD_PROPERTIES_FILE | Full name (including path) of the file containing all the system.* properties passed to the build. "system." prefix stripped off. The file uses Java properties file format (for example, special symbols are backslash-escaped).|
| TEAMCITY_VERSION |The version of TeamCity server. This property can be used to determine the build is run within TeamCity.|
| TEAMCITY_PROJECT_NAME| The name of the project the current build belongs to.|
| TEAMCITY_BUILDCONF_NAME | The name of the Build Configuration the current build belongs to.|
| BUILD_IS_PERSONAL|Is set to true if the build is a personal one. Is not defined otherwise.|
| BUILD_NUMBER | The build number assigned to the build by TeamCity using the build number format. The property is assigned based on the build number format.|

If we parsed the `TEAMCITY_BUILD_PROPERTIES_FILE` we could probably get allot more info from TeamCity

**Comment by mholo65** [comment]
ping @RichiCoder1 you had a Addin going and if I remember correctly, was about to submit a PR into Cake.Common for parsing the properties file.

**Comment by RichiCoder1** [comment]
Yup yup. Still very much WIP.

**Comment by devlead** [comment]
Example property file if someone wants something to parse in tests
```
#TeamCity build properties without 'system.' prefix
#Fri May 13 14:53:44 UTC 2016
agent.home.dir=C\:\\buildAgent
agent.name=JetBrains-dotnet-agent01
agent.ownPort=9090
agent.work.dir=Z\:\\TeamCityAgent\\work
build.number=680
build.vcs.number=a0a17ca98316e15a7866ccd4d8e34f57c3e4a4ad
build.vcs.number.1=a0a17ca98316e15a7866ccd4d8e34f57c3e4a4ad
build.vcs.number.Cake_HttpsGithubComCakeBuildCakeRefsHeadsDevelop=a0a17ca98316e15a7866ccd4d8e34f57c3e4a4ad
ec2.ami-id=ami-ba4e07d0
ec2.ami-launch-index=0
ec2.ami-manifest-path=(unknown)
ec2.instance-id=i-68c77dde
ec2.instance-type=c1.medium
ec2.local-hostname=ip-10-143-177-222.ec2.internal
ec2.local-ipv4=10.143.177.222
ec2.public-hostname=ec2-54-144-106-165.compute-1.amazonaws.com
ec2.public-ipv4=54.144.106.165
ec2.reservation-id=r-a6728a75
FxCopCmdFileVersion=1.35.60623.0
FxCopRoot=C\:\\Program Files (x86)\\Microsoft FxCop 1.35
java.io.tmpdir=Z\:\\TeamCityAgent\\temp\\buildTmp
Mono=C\:\\Program Files (x86)\\Mono-3.2.3\\bin\\mono.exe
Mono2.0_x86=C\:\\Program Files (x86)\\Mono-3.2.3\\lib\\mono\\2.0
Mono3.5_x86=C\:\\Program Files (x86)\\Mono-3.2.3\\lib\\mono\\3.5
Mono4.0_x86=C\:\\Program Files (x86)\\Mono-3.2.3\\lib\\mono\\4.0
Mono4.5_x86=C\:\\Program Files (x86)\\Mono-3.2.3\\lib\\mono\\4.5
MonoVersion=3.2.3
ruby.installed=true
teamcity.agent.cpuBenchmark=345
teamcity.agent.dotnet.agent_url=http\://localhost\:9090/RPC2
teamcity.agent.dotnet.build_id=217388
teamcity.auth.password=*redacted*
teamcity.auth.userId=*redacted*
teamcity.build.changedFiles.file=Z\:\\TeamCityAgent\\temp\\buildTmp\\changedFiles8578230850117063324.txt
teamcity.build.checkoutDir=Z\:\\TeamCityAgent\\work\\93195cd5a0cab8
teamcity.build.id=217388
teamcity.build.properties.file=Z\:\\TeamCityAgent\\temp\\buildTmp\\teamcity.build3718849549039712185.properties
teamcity.build.tempDir=Z\:\\TeamCityAgent\\temp\\buildTmp
teamcity.build.workingDir=Z\:\\TeamCityAgent\\work\\93195cd5a0cab8
teamcity.buildConfName=Cake Develop
teamcity.buildType.id=Cake_CakeMaster
teamcity.configuration.properties.file=Z\:\\TeamCityAgent\\temp\\buildTmp\\teamcity.config5859417396764621032.properties
teamcity.dotnet.nunitaddin=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.TeamCity.NUnitAddin-NUnit
teamcity.dotnet.nunitlauncher=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.BuildServer.NUnitLauncher.exe
teamcity.dotnet.nunitlauncher.msbuild.task=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.BuildServer.MSBuildLoggers.dll
teamcity.dotnet.nunitlauncher1.1=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.BuildServer.NUnitLauncher1.1.exe
teamcity.dotnet.nunitlauncher2.0=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.BuildServer.NUnitLauncher2.0.exe
teamcity.dotnet.nunitlauncher2.0.vsts=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.BuildServer.NUnitLauncher2.0.VSTS.exe
teamcity.dotnet.platform=C\:\\buildAgent\\plugins\\dotnetPlugin\\bin\\JetBrains.TeamCity.PlatformProcessRunner.1.1.exe
teamcity.nuget.feed.auth.serverRootUrlBased.server=http\://teamcity.codebetter.com/httpAuth/app/nuget/v1/FeedService.svc
teamcity.projectName=Cake
teamcity.runner.properties.file=Z\:\\TeamCityAgent\\temp\\buildTmp\\teamcity.runner312482885273880491.properties
teamcity.tests.recentlyFailedTests.file=Z\:\\TeamCityAgent\\temp\\buildTmp\\testsToRunFirst3236795199492064268.txt
teamcity.version=9.1.7 (build 37573)
```

**Comment by mholo65** [comment]
..And here's the grammar https://gist.github.com/RichiCoder1/ed5e1cdde18a5da2dbd115cd7e448a5d

**Comment by roly445** [comment]
Does anyone know if there has been any progress with this one.  It would be great if if I could get the build number into cake

**Comment by mholo65** [comment]
@roly445 reading value of environment variable BUILD_NUMBER. http://cakebuild.net/api/cake.common/7bef281e/3aa917af

```var buildnumber = EnvironmentVariable("BUILD_NUMBER");```

**Comment by pitermarx** [comment]
this is what I love about cake
```
string TeamcityInfo(string key)
{
    var file = EnvironmentVariable("TEAMCITY_BUILD_PROPERTIES_FILE"); 
    if (string.IsNullOrEmpty(file))
    {
        Error("Environment variable TEAMCITY_BUILD_PROPERTIES_FILE not found");
        return null;
    }
    var v = XmlPeek(file, "/properties/entry[@key='" + key + "']/text()", new XmlPeekSettings { DtdProcessing	= XmlDtdProcessing.Ignore });
    Verbose("{0} is {1}", key, v);
    return v;
}
```

**Comment by aateeque** [comment]
Pr #1433 provides this info in `ITeamCityProvider`

**Comment by pitermarx** [comment]
sure, but it does not provide access to all the properties in teamcity.
Should this issue be closed?

**Comment by mholo65** [comment]
@pitermax wuut? Is properties file available as XML? Last time I checked, it was a Java-properties file (similar to INI) just like the example @devlead provided above.

**Comment by Sam13** [comment]
Flag whether build is personal or not would be great to have...

**Comment by aateeque** [comment]
The `ITeamCityProvider` does not read the properties file; @Sam13 do you know where this flag is?

**Comment by Sam13** [comment]
There is a system property called `system.build.is.personal` which is set to `true` when build is personal.
Also there is an environment variable called `BUILD_IS_PERSONAL` which is also set to `true`.

I'm currently using my own custom workaround to check for the environment variable.

**Comment by aateeque** [comment]
Should probably add in `TeamCityBuildInfo` by reading env var...feel free to add otherwise I'll add at some point next week.

@gep13 should we close this issue; fixed with #1433
